modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C

```diff
@@ -0,0 +1,83 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "OPInterfaceBarrierMaterial.h"
+
+template<>
+InputParameters validParams<OPInterfaceBarrierMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.addParam<std::string>("function_name", "g", "actual name for g(eta_i)");
+  MooseEnum h_order("SIMPLE=0", "SIMPLE");
+  params.addParam<MooseEnum>("g_order", h_order, "Polynomial order of the switching function h(eta)");
+  params.addRequiredCoupledVar("etas", "eta_i order parameters, one for each h");
+
+  return params;
+}
+
+OPInterfaceBarrierMaterial::OPInterfaceBarrierMaterial(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    _function_name(getParam<std::string>("function_name")),
+    _g_order(getParam<MooseEnum>("g_order")),
+    _num_eta(coupledComponents("etas")),
+    _eta(_num_eta),
+    _prop_g(declareProperty<Real>(_function_name)),
+    _prop_dg(_num_eta),
+    _prop_d2g(_num_eta)
+
+{
+
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    _prop_d2g[i].resize(_num_eta);
+
+  // declare derivative properties, fetch eta values
+  std::vector<std::string> eta_name(_num_eta);
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    eta_name[i] = getVar("etas", i)->name();
+
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+    _prop_dg[i] = &declarePropertyDerivative<Real>(_function_name, eta_name[i]);
+    _eta[i] = &coupledValue("etas", i);
+    for (unsigned int j = i; j < _num_eta; ++j)
+    {
+      _prop_d2g[i][j] =
+      _prop_d2g[j][i] = &declarePropertyDerivative<Real>(_function_name, eta_name[i], eta_name[j]);
+    }
+  }
+}
+
+void
+OPInterfaceBarrierMaterial::computeQpProperties()
+{
+  // Initialize properties to zero before accumulating
+  _prop_g[_qp] = 0.0;
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+    (*_prop_dg[i])[_qp] = 0.0;
+    for (unsigned int j = i; j < _num_eta; ++j)
+      (*_prop_d2g[i][j])[_qp] = 0.0;
+  }
+
+  // Create Interface barrier preventing interfaces involving more than two order parameters
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = 0; j < _num_eta; ++j)
+    {
+      if (i == j) continue;
+      for (unsigned int k = 0; k < _num_eta; ++k)
+      {
+        if ((i == k) || (j == k)) continue;
+        switch (_g_order)
+        {
+          case 0: // SIMPLE
+            _prop_g[_qp]         +=  (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
```

Comment:
You are right! That is fine. I forgot about that external prefactor.

On Tue, Mar 15, 2016, 7:48 AM jbair34 <notifications@github.com> wrote:

> In modules/phase_field/src/materials/OPInterfaceBarrierMaterial.C
> <https://github.com/idaholab/moose/pull/6460#discussion_r56165145>:
>
> > +    for (unsigned int j = i; j < _num_eta; ++j)
> > +      (*_prop_d2g[i][j])[_qp] = 0.0;
> > +  }
> > +
> > +  // Create Interface barrier preventing interfaces involving more than two order parameters
> > +  for (unsigned int i = 0; i < _num_eta; ++i)
> > +    for (unsigned int j = 0; j < _num_eta; ++j)
> > +    {
> > +      if (i == j) continue;
> > +      for (unsigned int k = 0; k < _num_eta; ++k)
> > +      {
> > +        if ((i == k) || (j == k)) continue;
> > +        switch (_g_order)
> > +        {
> > +          case 0: // SIMPLE
> > +            _prop_g[_qp]         +=  (*_eta[i])[_qp] * (*_eta[i])[_qp] * (*_eta[j])[_qp] * (*_eta[j])[_qp] * (*_eta[k])[_qp] * (*_eta[k])[_qp];
>
> So I specifically didn't add a prefactor because the
> MultiBarrierFunctionMaterial does not have one and the prefactor is
> controlled by the DerivativeMultiphaseMaterial anyways... Do you want me to
> add one? I thought it might be confusing to have two places to add the
> prefactor.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/idaholab/moose/pull/6460/files#r56165145
>